x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyExecutor.java

```diff
@@ -25,6 +27,7 @@
     private final IndexNameExpressionResolver indexNameExpressionResolver;
     private final LongSupplier nowSupplier;
     private final int fetchSize;
+    private final ConcurrentHashMap<String, Semaphore> policyLocks = new ConcurrentHashMap<>();
```

Comment:
Maybe uses `org.elasticsearch.common.util.concurrent.KeyedLock`? 
(this also uses a `ConcurrentMap`, but can be directly used out of the box)

Reply:
`KeyedLock` is implemented using reentrant locks, which retain a reference of the thread which currently holds the lock. The thread that holds the lock is allowed to acquire it again, by nature of it being reentrant. If a request to launch an enrich policy that is already in flight is serviced by a thread in the pool that held the lock from the first run, then it would be allowed to execute since the thread already has "ownership" of the lock.

Additionally, the `KeyedLock` must be released using the `Releasable` instance, which unlocks the lock and removes it from the backing map. The underlying `ReentrantLock` throws a monitor exception if it is unlocked by any thread other than the one that currently holds the lock. This leads to problem number 2, in that the Policy execution is asynchronous. When the operation concludes and the lock is released, it is unlikely(impossible?) that it will be done from the same thread that owns the lock.

Using a binary semaphore gives us mutual exclusion across multiple operations on the same thread, and freedom to discard the lock from any thread that handles the end of the operation.

Reply:
👍 Thanks for explaining why using `KeyedLock` here isn't a good choice.